File: 0518-coin-change-ii/0518-coin-change-ii.py
class Solution:
    def bottomUp(self, coins, target):
        n=len(coins)
        dp=[[0 for _ in range(target+1)]for _ in range(n)]
        
        for i in range(n):
            dp[i][0]=1
        for i in range(1,target+1):
            if i%coins[0]==0:
                dp[0][i]=1
        for i in range(1, n):
            coin = coins[i]
            for j in range(1, target + 1):
                if j - coin >= 0:
                    dp[i][j] = dp[i][j - coin] + dp[i - 1][j]
                else:
                    dp[i][j] = dp[i - 1][j]
        
        return dp[-1][-1]
                
        
    def change(self, amount: int, coins: List[int]) -> int:
        return self.bottomUp(coins, amount)
        
        

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int dfs(int amount, vector<int>& coins,int idx,vector<vector<int>>&dp)
    {
        if(idx==coins.size())
        {
            return 0;
        }
        
        if(amount == 0)
        {
            return 1;
        }
        if(dp[idx][amount]!=-1)
        {
            return dp[idx][amount];
        }
        int take=0,dont_take=0;
        if(coins[idx]<=amount)
        {
            take=dfs(amount-coins[idx],coins,idx,dp);
        }
        dont_take=dfs(amount,coins,idx+1,dp);
        
        return dp[idx][amount]=take+dont_take;
    }
    int change(int amt, vector<int>& coins) {
        int n=coins.size();
        
        vector<vector<int>>dp(n+1,vector<int>(amt+1,-1));
        
        return dfs(amt,coins,0,dp);
    }
};